{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "87d9970d-0d30-4488-b896-086215563a8f",
   "metadata": {},
   "source": [
    "# 1. Downloading and assembling microbial sequence data\n",
    "\n",
    "Welcome to the **Downloading and assembling microbial sequence data** supplementary materials. This document is written as a [Jupyter](https://jupyter.org/) notebook, which means that it contains both written documentation as well as runnable code in the same document.\n",
    "\n",
    "There are two methods you can use to run the code here: (1) either copy-paste the bash commands to a separate terminal, or (2) you can run directly in Juptyer."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b48d0acf-c235-4087-9834-8ebd2cbc51e1",
   "metadata": {},
   "source": [
    "## 1.1. Copy-paste commands to a separate terminal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe1a4579-8c0d-4ff7-a7d4-3cc4dc00756b",
   "metadata": {},
   "source": [
    "## 1.2. Running from within Jupyter\n",
    "\n",
    "The easiest way to run the code is to click the **Play** button in the Jupyter notebook as shown below:\n",
    "\n",
    "![run-jupyter.png](images/run-jupyter.png)\n",
    "\n",
    "This will run a single cell (block of code or text) and advance to the next cell.\n",
    "\n",
    "If you, instead, which to run everything in this notebook all at once you can find the **Kerenel** option in the menu at the top and select **Restart Kerenel and Run All Cells..**, which will run everything in this notebook all at once.\n",
    "\n",
    "![restart-kernel-jupyter.png](images/restart-kernel-jupyter.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2160c6a-6691-44ad-a8d4-ad0dc8322d14",
   "metadata": {},
   "source": [
    "# 2. Download genomes from NCBI\n",
    "\n",
    "In this section we will go over how to download genomes from [NCBI's Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra/) by using the [SRA Tools](https://github.com/ncbi/sra-tools) command-line software provided by NCBI."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccdc0ad5-173a-4271-9a24-513565b06c35",
   "metadata": {},
   "source": [
    "## 2.1. Install `sra-tools` using conda\n",
    "\n",
    "This will create a new conda environment `sra-tools` and install the package `sra-tools` in this environment.\n",
    "\n",
    "The `-y` option here means to automatically answer `yes` to any prompts for input by conda. This is important when running using Jupyter (but can be left out if you are running via the command-line)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8c678e8d-aa50-4f3b-b393-92026b899c2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/sra-tools\n",
      "\n",
      "  added / updated specs:\n",
      "    - sra-tools\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
      "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu\n",
      "  c-ares             conda-forge/linux-64::c-ares-1.18.1-h7f98852_0\n",
      "  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0\n",
      "  curl               conda-forge/linux-64::curl-7.81.0-h2574ce0_0\n",
      "  hdf5               conda-forge/linux-64::hdf5-1.10.6-nompi_h6a2412b_1114\n",
      "  icu                conda-forge/linux-64::icu-69.1-h9c3ff4c_0\n",
      "  krb5               conda-forge/linux-64::krb5-1.19.2-hcc1bbae_3\n",
      "  libcurl            conda-forge/linux-64::libcurl-7.81.0-h2574ce0_0\n",
      "  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2\n",
      "  libev              conda-forge/linux-64::libev-4.33-h516909a_1\n",
      "  libgcc-ng          conda-forge/linux-64::libgcc-ng-11.2.0-h1d223b6_11\n",
      "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-11.2.0-h69a702a_11\n",
      "  libgfortran5       conda-forge/linux-64::libgfortran5-11.2.0-h5c6108e_11\n",
      "  libgomp            conda-forge/linux-64::libgomp-11.2.0-h1d223b6_11\n",
      "  libiconv           conda-forge/linux-64::libiconv-1.16-h516909a_0\n",
      "  libnghttp2         conda-forge/linux-64::libnghttp2-1.43.0-h812cca2_1\n",
      "  libssh2            conda-forge/linux-64::libssh2-1.10.0-ha56f1ee_2\n",
      "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-11.2.0-he4da1e4_11\n",
      "  libxml2            conda-forge/linux-64::libxml2-2.9.12-h885dcf4_1\n",
      "  libzlib            conda-forge/linux-64::libzlib-1.2.11-h36c2ea0_1013\n",
      "  ncbi-ngs-sdk       bioconda/linux-64::ncbi-ngs-sdk-2.11.2-hff44eed_0\n",
      "  ncurses            conda-forge/linux-64::ncurses-6.3-h9c3ff4c_0\n",
      "  openssl            conda-forge/linux-64::openssl-1.1.1l-h7f98852_0\n",
      "  ossuuid            conda-forge/linux-64::ossuuid-1.6.2-hf484d3e_1000\n",
      "  perl               conda-forge/linux-64::perl-5.26.2-h36c2ea0_1008\n",
      "  perl-app-cpanminus bioconda/linux-64::perl-app-cpanminus-1.7044-pl526_1\n",
      "  perl-business-isbn bioconda/linux-64::perl-business-isbn-3.004-pl526_0\n",
      "  perl-business-isb~ bioconda/linux-64::perl-business-isbn-data-20140910.003-pl526_0\n",
      "  perl-carp          bioconda/linux-64::perl-carp-1.38-pl526_3\n",
      "  perl-constant      bioconda/linux-64::perl-constant-1.33-pl526_1\n",
      "  perl-data-dumper   bioconda/linux-64::perl-data-dumper-2.173-pl526_0\n",
      "  perl-encode        bioconda/linux-64::perl-encode-2.88-pl526_1\n",
      "  perl-exporter      bioconda/linux-64::perl-exporter-5.72-pl526_1\n",
      "  perl-extutils-mak~ bioconda/linux-64::perl-extutils-makemaker-7.36-pl526_1\n",
      "  perl-file-path     bioconda/linux-64::perl-file-path-2.16-pl526_0\n",
      "  perl-file-temp     bioconda/linux-64::perl-file-temp-0.2304-pl526_2\n",
      "  perl-mime-base64   bioconda/linux-64::perl-mime-base64-3.15-pl526_1\n",
      "  perl-parent        bioconda/linux-64::perl-parent-0.236-pl526_1\n",
      "  perl-uri           bioconda/linux-64::perl-uri-1.76-pl526_0\n",
      "  perl-xml-libxml    bioconda/linux-64::perl-xml-libxml-2.0132-pl526h7ec2d77_1\n",
      "  perl-xml-namespac~ bioconda/linux-64::perl-xml-namespacesupport-1.12-pl526_0\n",
      "  perl-xml-sax       bioconda/noarch::perl-xml-sax-1.02-pl526_0\n",
      "  perl-xml-sax-base  bioconda/linux-64::perl-xml-sax-base-1.09-pl526_0\n",
      "  perl-xsloader      bioconda/linux-64::perl-xsloader-0.24-pl526_0\n",
      "  sra-tools          bioconda/linux-64::sra-tools-2.11.0-pl5262h314213e_1\n",
      "  tk                 conda-forge/linux-64::tk-8.6.11-h27826a3_1\n",
      "  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1\n",
      "  zlib               conda-forge/linux-64::zlib-1.2.11-h36c2ea0_1013\n",
      "\n",
      "\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate sra-tools\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda create -y -n sra-tools sra-tools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ee7570f-c35f-462c-9979-daa0999b01f5",
   "metadata": {},
   "source": [
    "## 2.2. Verify `sra-tools` is working\n",
    "\n",
    "Let's verify that the tools found in the package `sra-tools` are working properly. We can do this by running one of the installed commands, `prefetch` like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b2c752f6-108e-4ff9-8538-c49e3177f415",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?2004l\n",
      "Usage: prefetch [ options ] [ accessions(s)... ]\n",
      "\n",
      "Parameters:\n",
      "\n",
      "  accessions(s)                    list of accessions to process\n",
      "\n",
      "\n",
      "Options:\n",
      "\n",
      "  -T|--type <file-type>            Specify file type to download. Default: sra\n",
      "  -N|--min-size <size>             Minimum file size to download in KB\n",
      "                                     (inclusive).\n",
      "  -X|--max-size <size>             Maximum file size to download in KB\n",
      "                                     (exclusive). Default: 20G\n",
      "  -f|--force <no|yes|all|ALL>      Force object download - one of: no, yes,\n",
      "                                     all, ALL. no [default]: skip download if\n",
      "                                     the object if found and complete; yes:\n",
      "                                     download it even if it is found and is\n",
      "                                     complete; all: ignore lock files (stale\n",
      "                                     locks or it is being downloaded by\n",
      "                                     another process - use at your own\n",
      "                                     risk!); ALL: ignore lock files, restart\n",
      "                                     download from beginning\n",
      "  -p|--progress                    Show progress\n",
      "  -r|--resume <yes|no>             Resume partial downloads - one of: no, yes\n",
      "                                     [default]\n",
      "  -C|--verify <yes|no>             Verify after download - one of: no, yes\n",
      "                                     [default]\n",
      "  -c|--check-all                   Double-check all refseqs\n",
      "  -o|--output-file <file>          Write file to <file> when downloading\n",
      "                                     single file\n",
      "  -O|--output-directory <directory>\n",
      "                                   Save files to <directory>/\n",
      "     --ngc <path>                  <path> to ngc file\n",
      "     --perm <path>                 <path> to permission file\n",
      "     --location <location>         location in cloud\n",
      "     --cart <path>                 <path> to cart file\n",
      "  -V|--version                     Display the version of the program\n",
      "  -v|--verbose                     Increase the verbosity of the program\n",
      "                                     status messages. Use multiple times for\n",
      "                                     more verbosity.\n",
      "  -L|--log-level <level>           Logging level as number or enum string.\n",
      "                                     One of\n",
      "                                     (fatal|sys|int|err|warn|info|debug) or\n",
      "                                     (0-6) Current/default is warn\n",
      "     --option-file file            Read more options and parameters from the\n",
      "                                     file.\n",
      "  -h|--help                        print this message\n",
      "\n",
      "\"prefetch\" version 2.11.0\n",
      "\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda run -n sra-tools prefetch --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af63b69b-e8bd-4686-b818-9639e6edd259",
   "metadata": {},
   "source": [
    "Notice that I ran by first specifying `conda run -n sra-tools [...]`. The `conda run` part lets you specify a particular environment to activate first before trying to run a tool. It's useful in cases where you are installing each tool in a separate environment (as we are doing here)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8f42647-2ae8-47d5-96b7-67ec2e045a85",
   "metadata": {},
   "source": [
    "## 2.3. Prefetch genomes (`prefetch`)\n",
    "\n",
    "The [prefetch](https://github.com/ncbi/sra-tools/wiki/08.-prefetch-and-fasterq-dump) command that is part of the `sra-tools` package can be used to download and store sequence read data from NCBI's Sequence Read Archive. You can run it like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "222fa969-b9a7-4ef3-a550-088c040ed97c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?2004l\n",
      "2022-01-20T23:43:56 prefetch.2.11.0: 1) 'SRR3028792' is found locally\n",
      "2022-01-20T23:43:56 prefetch.2.11.0: 'SRR3028792' has 0 unresolved dependencies\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda run -n sra-tools prefetch SRR3028792"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5efaef5b-0d5a-4bb8-bcb7-cb053bf12bbc",
   "metadata": {},
   "source": [
    "You can even pass multiple accession numbers to `prefetch` like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "870dcff6-61d5-4c6f-b738-86f08e7cbe20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?2004l\n",
      "2022-01-20T23:44:30 prefetch.2.11.0: 1) 'SRR3028792' is found locally\n",
      "2022-01-20T23:44:30 prefetch.2.11.0: 'SRR3028792' has 0 unresolved dependencies\n",
      "\n",
      "2022-01-20T23:44:31 prefetch.2.11.0: 2) 'SRR3028793' is found locally\n",
      "2022-01-20T23:44:31 prefetch.2.11.0: 'SRR3028793' has 0 unresolved dependencies\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda run -n sra-tools prefetch SRR3028792 SRR3028793"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7271749-f8d7-4693-bac2-47f1289761f8",
   "metadata": {},
   "source": [
    "The `prefetch` command downloads the data from NCBI and stores on your computer in a special directory by default. You can see which directory using the `srapath` command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "07b56476-0e95-42f8-9d68-7b6d1c159355",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/CSCScience.ca/apetkau/ncbi/public/sra/SRR3028792.sra\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda run -n sra-tools srapath SRR3028792"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a50478-ec6b-4e49-895b-d6efdbfcc7d2",
   "metadata": {},
   "source": [
    "For how to know what identifier to use for `prefetch`, this can be found on the NCBI website.\n",
    "\n",
    "For example for `SRR3028792`, you can find this identifier (and the associated biological sample) at https://www.ncbi.nlm.nih.gov/sra/?term=SRR3028792\n",
    "\n",
    "![genome-sra-identifier.png](images/genome-sra-identifier.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06866e80-a81c-4a6b-bc6b-7d6d50eb5cc7",
   "metadata": {},
   "source": [
    "## 2.4. Convert to fastq (`fasterq-dump`)\n",
    "\n",
    "Once you've prefetched the sequence data you can convert to the **fastq** format using `fasterq-dump` (which is named based on the older `fastq-dump` tool)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "41443e95-f5d0-4cec-96b9-5759213f3b28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spots read      : 824,262\n",
      "reads read      : 1,648,524\n",
      "reads written   : 1,648,524\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda run -n sra-tools fasterq-dump SRR3028792"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f504e35c-005e-4d6a-81e3-de6ed3a862ee",
   "metadata": {},
   "source": [
    "This will write two `*.fastq` files do your current directory which contain the forward (`_1.fastq`) and reverse (`_2.fastq`) reads.\n",
    "\n",
    "We can use the `ls` command to **l**i**s**t the files and show additional information (such as file size)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fcd6f919-2012-4c4d-9548-c20997eca57a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 apetkau grp_apetkau 398M Jan 20 17:54 SRR3028792_2.fastq\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 396M Jan 20 17:54 SRR3028792_1.fastq\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "ls -lth *.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb39e0cb-669b-48bf-a09e-2666c9f75a86",
   "metadata": {},
   "source": [
    "The `396M` part shows us that `SRR3028792_1.fastq` is **396 MB**. We can compress these files to save on storage space using the `gzip` command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "27e0db23-e1f0-4c6e-ab2a-a196b4cc2222",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "gzip *.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b2d0bcf6-8489-4cee-94bb-3e2a5207b4a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 apetkau grp_apetkau 138M Jan 20 17:54 \u001b[0m\u001b[01;31mSRR3028792_2.fastq.gz\u001b[0m\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 110M Jan 20 17:54 \u001b[01;31mSRR3028792_1.fastq.gz\u001b[0m\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "ls -lth *.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d51cbd5-cca1-4658-a997-9c423e35a7fb",
   "metadata": {},
   "source": [
    "Now the files end in `.fastq.gz` instead of `.fastq` and are much smaller (`SRR3028792_1.fastq.gz` is **110 MB** compared to **396 MB** for the uncompressed file)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1becc05-d400-4700-9af9-50d1a75b0f8f",
   "metadata": {},
   "source": [
    "We can also look at the file contents using the `cat` (or `zcat` for files ending in `.gz`) commands along with `head`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2b30427f-5412-4985-b1a0-6b22ff28a0be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR3028792.1 1 length=90\n",
      "CTTCATAATCAGGCGATAAATGCCCACCACTTTAGGCTTTCTGGCGCGGATAGCCTCCCCAATAAAATCTTTACGCGTACGCTTTGCTTC\n",
      "+SRR3028792.1 1 length=90\n",
      "AABC-C--CE,-,C++@,,,<,CCC,BB,:CCE,,,;CCCCC,,6+,+++8,,:,99,,8,,:,,,,<CCEE9C,@+8+8++88C,9CCC\n",
      "@SRR3028792.2 2 length=271\n",
      "ATACACGCCAGCGCGAATATTTATCGCTAGCATCAGGAAAACCAACGAAACAGCCCTCGTTGATTTGCCCCCTGAACAGTGGGCTTTTTTCGCGTACAGATGGCTCCGGCATTAGACCTGTGACATCCCGATACTATTCACATTCCGCAGCTCATACTTCACTTCTGCCCGTAATACGCCTTCGGACCATGCTTACGCATGACGTGTTTATTCATCAGATCATCGTCGATGGGATTCAGCGCCGGGTTAATGCCGCGCGCAATCCACGCCA\n",
      "+SRR3028792.2 2 length=271\n",
      "-86A<F-@F7+C+C@77C,CFF9EE8FE7+C,CE,,,,,,,;@,,8,,,,+,,9:,9,+::,,CEE,CEEDEF7,,:,,,,,,:CEFFEB=+8+++++,,:,,49A?++++8B,,,:??,:,,:,CEFF7+8++:,:A@,>,DDDD,+++88@,@,@CCC,@@B>@,>@D*@**6,5*4*6>?***>:*<,5=>,>*;5=*,,*2*<:C+<CF7A?+++3**;:*9C))7+***29F**9)7<5))8>)*9*8;5;)7)7((-57)5(4:(\n",
      "@SRR3028792.3 3 length=181\n",
      "TGTCGAGATGGTGGCTGAACTGCATCGGGTCTATAGCGCCCAGACTCAGGCGTATGTCCTGTTTAATGAAGAGAGTGCACTTTCGCAAGCGCTTCTGTTGCAACGTCATGGGGAAGAAGATTTTCTTGCCTTTCTCCGGGCGGCGACCGCACCCCTGCGTCACGCACTGTCGACATATCCC\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "zcat SRR3028792_1.fastq.gz | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ca4e652-f1c4-4a8c-bbfa-8380a0fa9713",
   "metadata": {},
   "source": [
    "The command `zcat` uncompresses any `*.gz` files and then prints out the contents. Since these are large files we likely only want to look at the first few lines in the file. The command `head` will print out the first 10 lines by default (and ignore all the rest). The `|` character tells bash that we want to take the output of `zcat SRR3028792_1.fastq.gz` (which will print all the lines of a file) and forward it into the `head` command (which will print out the first 10 lines of data and ignore the rest).\n",
    "\n",
    "The end result is we see the first 10 lines of the `SRR3028792_1.fastq.gz` file.\n",
    "\n",
    "We can do the same thing for the reverse reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "69c3e30d-ebae-4711-aad2-33ee7dc1ace7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR3028792.1 1 length=300\n",
      "CTCTCAAACCTTCCTCGTTACTTTTTTCTTTCCGCCTCTCTCCTCCCCCCCCCGCCTCCCCTCTTTTCCCTTTCTCTCCTGCTTCTGCCCCTCTCTCTTCTCCTCCTCTGCCTCTTCCTCCTCCCCCTTACTTCTTCCTCTCTTTCCTCCCCCTCTCCTTCCTCCTCCCCCCCTCTCCCTCCCCCCTTCCCTCCCCTCTCCCTCCCCCCCCTCCCCCCCCCTCCCTCCCTCCTCCCCCCTCTCTCTCCCCTCTCTCCCTCCCCCCTCCCCCCCCTCTCCTCCCCCCCCCCCCTCCCCCCC\n",
      "+SRR3028792.1 1 length=300\n",
      "--,-8,-,;:,;,;,;,:,,,,,6;;@,<@,;;,,8,8;,66;;;,88++++++678,,,,6,,6,,,6,9:?,59,:95,,:9,:,,,,9:+995:9:,:,9,4,999,,9,94,99,,8,,,++84,,,,3,8,,,,,63,,3,,33,+++++,6,,,,,,,,,,,+++*******0******))))*)+*))))))0)))))))))(((((((((((,((,((,(((((((((((((,(())))(,((((.))(((((((((((((((((((,((()((((((((,((((,,(((((\n",
      "@SRR3028792.2 2 length=300\n",
      "TTGCTTCGATTTCTCTCTCCCTTCACCCGGCGCTGACTCCCCTCTCCTCTTCTCTGATGCCTCCTCACTTCATCCTTCCTCCTCCTCCCCCTTCGTATTACTTTCCTCCGTCCCTTCTGCTCTCCTCCCTCTCCCTCTTTTCTCTCTTTCACATTTCTACTTCCGCCCCCCTCTTTCCCCCCACCCACCCCCCTTTTCCGCGTGCCTCTCCACTCTCGCTTTTTCCTTCCTTCTCCTGTTTCTCCCTCTCACTCTTCCCTCTTGCTTTTTTCTTTCTCTTCTACCCCTCTGCCCCTTCCC\n",
      "+SRR3028792.2 2 length=300\n",
      "--,8,;-+-<;,<,C,;,,6,6;,,;6@++8+88++,;C@C,;B,,;,,<;,<;;,,<,,,<,,,:,:C9E,<,:,9,,,<,:,,9:@,,+7,:,:,:9,:,,,9,,,,+9,+,,9,9,,,9499,,,+9,9,,,9,99,49,,,,6,44,7,,,47=,,663=++++6=+66+6=D?+5++++****33903/*333*)))))0)+)1/()/1(/)(((//./.)75)(1.)/7@9))/)11)).((.()).)464).).,)).).),)(,))-.-.65)6)--(,((()).,-((..(\n",
      "@SRR3028792.3 3 length=300\n",
      "CTTCTATTTCTCCCGTTCCTCACTCCGCCTTCCGGTCCCCCCCCGTCTTCACTCAATAACATCTTCTTCCCCCTCTCTTTCCACCCTCACCGCTTCCGCCCCTTCTCTCTCTTCCTTCACCTTCCCCTCCGCCTTCTTCTTCTCCCTCTCTACCCCATCCCTTTCAGCCCCCATCTCTCCTCTCTCTCTTCTCCCCCTCTTCCGCTTCCCCCCACTTCTTCTCTCTCCTTCTCTCTCTTCCCCCTCTCCTTTCTCTCCTCTCTTCCCTCCCCTCCCCTCCCCTCCCTCCTCCTCTCCTCC\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "zcat SRR3028792_2.fastq.gz | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99d8fe83-bf09-41c5-9bc6-0979bf4ce00e",
   "metadata": {},
   "source": [
    "# 3. Quality filter files (`fastp`)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd71b01d-f656-4cd6-91f2-e7dc66c05db6",
   "metadata": {},
   "source": [
    "## 3.1. Install `fastp` using conda"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "faa459cb-e786-4bdf-9f48-d86b734432fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/fastp\n",
      "\n",
      "  added / updated specs:\n",
      "    - fastp\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
      "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu\n",
      "  fastp              bioconda/linux-64::fastp-0.23.2-hd36eab0_1\n",
      "  isa-l              conda-forge/linux-64::isa-l-2.30.0-ha770c72_4\n",
      "  libdeflate         conda-forge/linux-64::libdeflate-1.9-h7f98852_0\n",
      "  libgcc-ng          conda-forge/linux-64::libgcc-ng-11.2.0-h1d223b6_11\n",
      "  libgomp            conda-forge/linux-64::libgomp-11.2.0-h1d223b6_11\n",
      "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-11.2.0-he4da1e4_11\n",
      "\n",
      "\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate fastp\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda create -y -n fastp fastp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7317080a-1fea-4989-9516-174e9cf0d051",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: fastp [options] ... \n",
      "options:\n",
      "  -i, --in1                            read1 input file name (string [=])\n",
      "  -o, --out1                           read1 output file name (string [=])\n",
      "  -I, --in2                            read2 input file name (string [=])\n",
      "  -O, --out2                           read2 output file name (string [=])\n",
      "      --unpaired1                      for PE input, if read1 passed QC but read2 not, it will be written to unpaired1. Default is to discard it. (string [=])\n",
      "      --unpaired2                      for PE input, if read2 passed QC but read1 not, it will be written to unpaired2. If --unpaired2 is same as --unpaired1 (default mode), both unpaired reads will be written to this same file. (string [=])\n",
      "      --overlapped_out                 for each read pair, output the overlapped region if it has no any mismatched base. (string [=])\n",
      "      --failed_out                     specify the file to store reads that cannot pass the filters. (string [=])\n",
      "  -m, --merge                          for paired-end input, merge each pair of reads into a single read if they are overlapped. The merged reads will be written to the file given by --merged_out, the unmerged reads will be written to the files specified by --out1 and --out2. The merging mode is disabled by default.\n",
      "      --merged_out                     in the merging mode, specify the file name to store merged output, or specify --stdout to stream the merged output (string [=])\n",
      "      --include_unmerged               in the merging mode, write the unmerged or unpaired reads to the file specified by --merge. Disabled by default.\n",
      "  -6, --phred64                        indicate the input is using phred64 scoring (it'll be converted to phred33, so the output will still be phred33)\n",
      "  -z, --compression                    compression level for gzip output (1 ~ 9). 1 is fastest, 9 is smallest, default is 4. (int [=4])\n",
      "      --stdin                          input from STDIN. If the STDIN is interleaved paired-end FASTQ, please also add --interleaved_in.\n",
      "      --stdout                         stream passing-filters reads to STDOUT. This option will result in interleaved FASTQ output for paired-end output. Disabled by default.\n",
      "      --interleaved_in                 indicate that <in1> is an interleaved FASTQ which contains both read1 and read2. Disabled by default.\n",
      "      --reads_to_process               specify how many reads/pairs to be processed. Default 0 means process all reads. (int [=0])\n",
      "      --dont_overwrite                 don't overwrite existing files. Overwritting is allowed by default.\n",
      "      --fix_mgi_id                     the MGI FASTQ ID format is not compatible with many BAM operation tools, enable this option to fix it.\n",
      "  -V, --verbose                        output verbose log information (i.e. when every 1M reads are processed).\n",
      "  -A, --disable_adapter_trimming       adapter trimming is enabled by default. If this option is specified, adapter trimming is disabled\n",
      "  -a, --adapter_sequence               the adapter for read1. For SE data, if not specified, the adapter will be auto-detected. For PE data, this is used if R1/R2 are found not overlapped. (string [=auto])\n",
      "      --adapter_sequence_r2            the adapter for read2 (PE data only). This is used if R1/R2 are found not overlapped. If not specified, it will be the same as <adapter_sequence> (string [=auto])\n",
      "      --adapter_fasta                  specify a FASTA file to trim both read1 and read2 (if PE) by all the sequences in this FASTA file (string [=])\n",
      "      --detect_adapter_for_pe          by default, the auto-detection for adapter is for SE data input only, turn on this option to enable it for PE data.\n",
      "  -f, --trim_front1                    trimming how many bases in front for read1, default is 0 (int [=0])\n",
      "  -t, --trim_tail1                     trimming how many bases in tail for read1, default is 0 (int [=0])\n",
      "  -b, --max_len1                       if read1 is longer than max_len1, then trim read1 at its tail to make it as long as max_len1. Default 0 means no limitation (int [=0])\n",
      "  -F, --trim_front2                    trimming how many bases in front for read2. If it's not specified, it will follow read1's settings (int [=0])\n",
      "  -T, --trim_tail2                     trimming how many bases in tail for read2. If it's not specified, it will follow read1's settings (int [=0])\n",
      "  -B, --max_len2                       if read2 is longer than max_len2, then trim read2 at its tail to make it as long as max_len2. Default 0 means no limitation. If it's not specified, it will follow read1's settings (int [=0])\n",
      "  -D, --dedup                          enable deduplication to drop the duplicated reads/pairs\n",
      "      --dup_calc_accuracy              accuracy level to calculate duplication (1~6), higher level uses more memory (1G, 2G, 4G, 8G, 16G, 24G). Default 1 for no-dedup mode, and 3 for dedup mode. (int [=0])\n",
      "      --dont_eval_duplication          don't evaluate duplication rate to save time and use less memory.\n",
      "  -g, --trim_poly_g                    force polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data\n",
      "      --poly_g_min_len                 the minimum length to detect polyG in the read tail. 10 by default. (int [=10])\n",
      "  -G, --disable_trim_poly_g            disable polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data\n",
      "  -x, --trim_poly_x                    enable polyX trimming in 3' ends.\n",
      "      --poly_x_min_len                 the minimum length to detect polyX in the read tail. 10 by default. (int [=10])\n",
      "  -5, --cut_front                      move a sliding window from front (5') to tail, drop the bases in the window if its mean quality < threshold, stop otherwise.\n",
      "  -3, --cut_tail                       move a sliding window from tail (3') to front, drop the bases in the window if its mean quality < threshold, stop otherwise.\n",
      "  -r, --cut_right                      move a sliding window from front to tail, if meet one window with mean quality < threshold, drop the bases in the window and the right part, and then stop.\n",
      "  -W, --cut_window_size                the window size option shared by cut_front, cut_tail or cut_sliding. Range: 1~1000, default: 4 (int [=4])\n",
      "  -M, --cut_mean_quality               the mean quality requirement option shared by cut_front, cut_tail or cut_sliding. Range: 1~36 default: 20 (Q20) (int [=20])\n",
      "      --cut_front_window_size          the window size option of cut_front, default to cut_window_size if not specified (int [=4])\n",
      "      --cut_front_mean_quality         the mean quality requirement option for cut_front, default to cut_mean_quality if not specified (int [=20])\n",
      "      --cut_tail_window_size           the window size option of cut_tail, default to cut_window_size if not specified (int [=4])\n",
      "      --cut_tail_mean_quality          the mean quality requirement option for cut_tail, default to cut_mean_quality if not specified (int [=20])\n",
      "      --cut_right_window_size          the window size option of cut_right, default to cut_window_size if not specified (int [=4])\n",
      "      --cut_right_mean_quality         the mean quality requirement option for cut_right, default to cut_mean_quality if not specified (int [=20])\n",
      "  -Q, --disable_quality_filtering      quality filtering is enabled by default. If this option is specified, quality filtering is disabled\n",
      "  -q, --qualified_quality_phred        the quality value that a base is qualified. Default 15 means phred quality >=Q15 is qualified. (int [=15])\n",
      "  -u, --unqualified_percent_limit      how many percents of bases are allowed to be unqualified (0~100). Default 40 means 40% (int [=40])\n",
      "  -n, --n_base_limit                   if one read's number of N base is >n_base_limit, then this read/pair is discarded. Default is 5 (int [=5])\n",
      "  -e, --average_qual                   if one read's average quality score <avg_qual, then this read/pair is discarded. Default 0 means no requirement (int [=0])\n",
      "  -L, --disable_length_filtering       length filtering is enabled by default. If this option is specified, length filtering is disabled\n",
      "  -l, --length_required                reads shorter than length_required will be discarded, default is 15. (int [=15])\n",
      "      --length_limit                   reads longer than length_limit will be discarded, default 0 means no limitation. (int [=0])\n",
      "  -y, --low_complexity_filter          enable low complexity filter. The complexity is defined as the percentage of base that is different from its next base (base[i] != base[i+1]).\n",
      "  -Y, --complexity_threshold           the threshold for low complexity filter (0~100). Default is 30, which means 30% complexity is required. (int [=30])\n",
      "      --filter_by_index1               specify a file contains a list of barcodes of index1 to be filtered out, one barcode per line (string [=])\n",
      "      --filter_by_index2               specify a file contains a list of barcodes of index2 to be filtered out, one barcode per line (string [=])\n",
      "      --filter_by_index_threshold      the allowed difference of index barcode for index filtering, default 0 means completely identical. (int [=0])\n",
      "  -c, --correction                     enable base correction in overlapped regions (only for PE data), default is disabled\n",
      "      --overlap_len_require            the minimum length to detect overlapped region of PE reads. This will affect overlap analysis based PE merge, adapter trimming and correction. 30 by default. (int [=30])\n",
      "      --overlap_diff_limit             the maximum number of mismatched bases to detect overlapped region of PE reads. This will affect overlap analysis based PE merge, adapter trimming and correction. 5 by default. (int [=5])\n",
      "      --overlap_diff_percent_limit     the maximum percentage of mismatched bases to detect overlapped region of PE reads. This will affect overlap analysis based PE merge, adapter trimming and correction. Default 20 means 20%. (int [=20])\n",
      "  -U, --umi                            enable unique molecular identifier (UMI) preprocessing\n",
      "      --umi_loc                        specify the location of UMI, can be (index1/index2/read1/read2/per_index/per_read, default is none (string [=])\n",
      "      --umi_len                        if the UMI is in read1/read2, its length should be provided (int [=0])\n",
      "      --umi_prefix                     if specified, an underline will be used to connect prefix and UMI (i.e. prefix=UMI, UMI=AATTCG, final=UMI_AATTCG). No prefix by default (string [=])\n",
      "      --umi_skip                       if the UMI is in read1/read2, fastp can skip several bases following UMI, default is 0 (int [=0])\n",
      "  -p, --overrepresentation_analysis    enable overrepresented sequence analysis.\n",
      "  -P, --overrepresentation_sampling    one in (--overrepresentation_sampling) reads will be computed for overrepresentation analysis (1~10000), smaller is slower, default is 20. (int [=20])\n",
      "  -j, --json                           the json format report file name (string [=fastp.json])\n",
      "  -h, --html                           the html format report file name (string [=fastp.html])\n",
      "  -R, --report_title                   should be quoted with ' or \", default is \"fastp report\" (string [=fastp report])\n",
      "  -w, --thread                         worker thread number, default is 3 (int [=3])\n",
      "  -s, --split                          split output by limiting total split file number with this option (2~999), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (int [=0])\n",
      "  -S, --split_by_lines                 split output by limiting lines of each file with this option(>=1000), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (long [=0])\n",
      "  -d, --split_prefix_digits            the digits for the sequential number padding (1~10), default is 4, so the filename will be padded as 0001.xxx, 0 to disable padding (int [=4])\n",
      "      --cut_by_quality5                DEPRECATED, use --cut_front instead.\n",
      "      --cut_by_quality3                DEPRECATED, use --cut_tail instead.\n",
      "      --cut_by_quality_aggressive      DEPRECATED, use --cut_right instead.\n",
      "      --discard_unmerged               DEPRECATED, no effect now, see the introduction for merging.\n",
      "  -?, --help                           print this message\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda run -n fastp fastp --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e5630fe-c25a-491d-b1c0-888866bcb880",
   "metadata": {},
   "source": [
    "## 3.2. Evaluate quality of sequence reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b7e7f612-fca1-4242-80b8-428e1940a64b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read1 before filtering:\n",
      "total reads: 824262\n",
      "total bases: 176567188\n",
      "Q20 bases: 171950867(97.3855%)\n",
      "Q30 bases: 163432314(92.561%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 824262\n",
      "total bases: 177556496\n",
      "Q20 bases: 153139830(86.2485%)\n",
      "Q30 bases: 134061609(75.5036%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 788353\n",
      "total bases: 166850080\n",
      "Q20 bases: 163779407(98.1596%)\n",
      "Q30 bases: 156905038(94.0395%)\n",
      "\n",
      "Read2 after filtering:\n",
      "total reads: 788353\n",
      "total bases: 166698912\n",
      "Q20 bases: 149106629(89.4467%)\n",
      "Q30 bases: 132133429(79.2647%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 1576706\n",
      "reads failed due to low quality: 71818\n",
      "reads failed due to too many N: 0\n",
      "reads failed due to too short: 0\n",
      "reads with adapter trimmed: 46646\n",
      "bases trimmed due to adapters: 1033979\n",
      "\n",
      "Duplication rate: 0.665322%\n",
      "\n",
      "Insert size peak (evaluated by paired-end reads): 35\n",
      "\n",
      "JSON report: SRR3028792.json\n",
      "HTML report: fastp.html\n",
      "\n",
      "fastp --in1 SRR3028792_1.fastq.gz --in2 SRR3028792_2.fastq.gz --out1 SRR3028792_1.fp.fastq.gz --out2 SRR3028792_2.fp.fastq.gz –-html SRR3028792.html --json SRR3028792.json \n",
      "fastp v0.23.2, time used: 7 seconds\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda run -n fastp fastp --in1 SRR3028792_1.fastq.gz --in2 SRR3028792_2.fastq.gz --out1 SRR3028792_1.fp.fastq.gz --out2 SRR3028792_2.fp.fastq.gz –-html SRR3028792.html --json SRR3028792.json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "000237d7-e164-41d4-87ec-fb6cfe8033c4",
   "metadata": {},
   "source": [
    "# 4. Genome assembly (`skesa`)\n",
    "\n",
    "## 4.1. Install skesa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "da879b03-092a-46d6-92eb-3ec9870f7b8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/skesa\n",
      "\n",
      "  added / updated specs:\n",
      "    - skesa\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
      "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu\n",
      "  boost              conda-forge/linux-64::boost-1.70.0-py38h9de70de_1\n",
      "  boost-cpp          conda-forge/linux-64::boost-cpp-1.70.0-h7b93d67_3\n",
      "  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4\n",
      "  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0\n",
      "  icu                conda-forge/linux-64::icu-67.1-he1b5a44_0\n",
      "  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.36.1-hea4e1c9_2\n",
      "  libblas            conda-forge/linux-64::libblas-3.9.0-13_linux64_openblas\n",
      "  libcblas           conda-forge/linux-64::libcblas-3.9.0-13_linux64_openblas\n",
      "  libffi             conda-forge/linux-64::libffi-3.4.2-h7f98852_5\n",
      "  libgcc-ng          conda-forge/linux-64::libgcc-ng-11.2.0-h1d223b6_11\n",
      "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-11.2.0-h69a702a_11\n",
      "  libgfortran5       conda-forge/linux-64::libgfortran5-11.2.0-h5c6108e_11\n",
      "  libgomp            conda-forge/linux-64::libgomp-11.2.0-h1d223b6_11\n",
      "  liblapack          conda-forge/linux-64::liblapack-3.9.0-13_linux64_openblas\n",
      "  libnsl             conda-forge/linux-64::libnsl-2.0.0-h7f98852_0\n",
      "  libopenblas        conda-forge/linux-64::libopenblas-0.3.18-pthreads_h8fe5266_0\n",
      "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-11.2.0-he4da1e4_11\n",
      "  libzlib            conda-forge/linux-64::libzlib-1.2.11-h36c2ea0_1013\n",
      "  lz4-c              conda-forge/linux-64::lz4-c-1.9.3-h9c3ff4c_1\n",
      "  ncurses            conda-forge/linux-64::ncurses-6.3-h9c3ff4c_0\n",
      "  numpy              conda-forge/linux-64::numpy-1.22.0-py38h6ae9a64_1\n",
      "  openssl            conda-forge/linux-64::openssl-3.0.0-h7f98852_2\n",
      "  pip                conda-forge/noarch::pip-21.3.1-pyhd8ed1ab_0\n",
      "  python             conda-forge/linux-64::python-3.8.12-hf930737_2_cpython\n",
      "  python_abi         conda-forge/linux-64::python_abi-3.8-2_cp38\n",
      "  readline           conda-forge/linux-64::readline-8.1-h46c0cb4_0\n",
      "  setuptools         conda-forge/linux-64::setuptools-60.5.0-py38h578d9bd_0\n",
      "  skesa              bioconda/linux-64::skesa-2.4.0-he1c1bb9_0\n",
      "  sqlite             conda-forge/linux-64::sqlite-3.37.0-h9cd32fc_0\n",
      "  tk                 conda-forge/linux-64::tk-8.6.11-h27826a3_1\n",
      "  wheel              conda-forge/noarch::wheel-0.37.1-pyhd8ed1ab_0\n",
      "  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1\n",
      "  zlib               conda-forge/linux-64::zlib-1.2.11-h36c2ea0_1013\n",
      "  zstd               conda-forge/linux-64::zstd-1.4.9-ha95c52a_0\n",
      "\n",
      "\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate skesa\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda create -y -n skesa skesa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63b3e0c9-5349-48a6-b6d0-42219db8cc82",
   "metadata": {},
   "source": [
    "## 4.2. Assemble genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "783bc907-e4e4-40cc-8190-fbe03af9d060",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skesa --reads SRR3028792_1.fp.fastq.gz,SRR3028792_2.fp.fastq.gz --contigs_out SRR3028792.fasta \n",
      "\n",
      "Total mates: 1576706 Paired reads: 788353\n",
      "Reads acquired in  14.206267s wall, 13.930000s user + 0.240000s system = 14.170000s CPU (99.7%)\n",
      "\n",
      "Kmer len: 19\n",
      "Raw kmers: 305168274 Memory needed (GB): 5.85923 Memory available (GB): 29.8726 1 cycle(s) will be performed\n",
      "Distinct kmers: 172827\n",
      "Kmer count in  3.040845s wall, 115.810000s user + 7.850000s system = 123.660000s CPU (4066.6%)\n",
      "Uniq kmers merging in  0.100942s wall, 0.070000s user + 0.260000s system = 0.330000s CPU (326.9%)\n",
      "Adapters: 0 Reads before: 1576706 Sequence before: 333548982 Reads after: 1576706 Sequence after: 333548982 Reads clipped: 0\n",
      "Adapters clipped in  3.148439s wall, 115.890000s user + 8.110000s system = 124.000000s CPU (3938.5%)\n",
      "\n",
      "Kmer len: 21\n",
      "Raw kmers: 302014862 Memory needed (GB): 5.79869 Memory available (GB): 29.783 1 cycle(s) will be performed\n",
      "Distinct kmers: 5886038\n",
      "Kmer count in  2.713511s wall, 113.830000s user + 7.760000s system = 121.590000s CPU (4480.9%)\n",
      "Uniq kmers merging in  0.473926s wall, 0.950000s user + 1.510000s system = 2.460000s CPU (519.1%)\n",
      "Kmers branching in  0.523914s wall, 20.070000s user + 0.000000s system = 20.070000s CPU (3830.8%)\n",
      "\n",
      "Average read length: 212\n",
      "Genome size estimate: 4388683\n",
      "\n",
      "Kmer: 21 Graph size: 5886038 Contigs in: 0\n",
      "Valley: 7\n",
      "\n",
      "Mark used kmers in  0.000003s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 0\n",
      "Fragments before: 1867 4803188\n",
      "Fragments after: 1645 4803188\n",
      "New seeds: 1344\n",
      "New seeds in  1.186548s wall, 57.130000s user + 0.440000s system = 57.570000s CPU (4851.9%)\n",
      "Fragments before: 2688 56448\n",
      "Fragments after: 2688 56448\n",
      "Connectors: 0 Extenders: 2688\n",
      "Connections and extensions in  0.114978s wall, 4.240000s user + 0.050000s system = 4.290000s CPU (3731.2%)\n",
      "Contigs out: 1344 Genome: 4756810 N50: 7553 L50: 187\n",
      "Assembled in  1.324830s wall, 61.390000s user + 0.490000s system = 61.880000s CPU (4670.8%)\n",
      "\n",
      "\n",
      "Kmer len: 173\n",
      "Raw kmers: 99928415 Memory needed (GB): 6.71519 Memory available (GB): 29.678 1 cycle(s) will be performed\n",
      "Distinct kmers: 5408481\n",
      "Kmer count in  2.318123s wall, 91.270000s user + 6.710000s system = 97.980000s CPU (4226.7%)\n",
      "Uniq kmers merging in  1.037581s wall, 1.610000s user + 2.870000s system = 4.480000s CPU (431.8%)\n",
      "\n",
      "Average count: 50.4151 Max kmer: 173\n",
      "Connected: 9206 ambiguously connected: 286 from 10000 mate pairs\n",
      "Connect pairs in  1.582214s wall, 74.180000s user + 0.060000s system = 74.240000s CPU (4692.2%)\n",
      "\n",
      "N50 for inserts: 363\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 311364\n",
      "Reads for connection: 158848\n",
      "Internal reads: 23033\n",
      "Reads cleaned in  0.602199s wall, 23.300000s user + 0.220000s system = 23.520000s CPU (3905.7%)\n",
      "\n",
      "Kmer len: 35\n",
      "Raw kmers: 59168172 Memory needed (GB): 1.70404 Memory available (GB): 29.7137 1 cycle(s) will be performed\n",
      "Distinct kmers: 1815510\n",
      "Kmer count in  0.806838s wall, 27.770000s user + 1.790000s system = 29.560000s CPU (3663.7%)\n",
      "Uniq kmers merging in  0.145764s wall, 0.430000s user + 0.090000s system = 0.520000s CPU (356.7%)\n",
      "Kmers branching in  0.176914s wall, 6.850000s user + 0.000000s system = 6.850000s CPU (3871.9%)\n",
      "Kmer: 35 Graph size: 1815510 Contigs in: 1344\n",
      "Valley: 18\n",
      "\n",
      "Mark used kmers in  0.090208s wall, 4.270000s user + 0.000000s system = 4.270000s CPU (4733.5%)\n",
      "Kmers in multiple/single contigs: 0 1419736\n",
      "Fragments before: 1765 106746\n",
      "Fragments after: 1736 106746\n",
      "New seeds: 62\n",
      "New seeds in  0.325714s wall, 13.350000s user + 0.010000s system = 13.360000s CPU (4101.8%)\n",
      "Fragments before: 2812 93582\n",
      "Fragments after: 1747 93582\n",
      "Connectors: 1065 Extenders: 682\n",
      "Connections and extensions in  0.165334s wall, 5.130000s user + 0.020000s system = 5.150000s CPU (3114.9%)\n",
      "Contigs out: 342 Genome: 4801583 N50: 39935 L50: 37\n",
      "Assembled in  0.595542s wall, 22.770000s user + 0.030000s system = 22.800000s CPU (3828.4%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 106218\n",
      "Reads for connection: 63592\n",
      "Internal reads: 6675\n",
      "Reads cleaned in  0.531295s wall, 14.950000s user + 0.090000s system = 15.040000s CPU (2830.8%)\n",
      "\n",
      "Kmer len: 51\n",
      "Raw kmers: 17693418 Memory needed (GB): 0.50957 Memory available (GB): 29.8143 1 cycle(s) will be performed\n",
      "Distinct kmers: 514372\n",
      "Kmer count in  0.227535s wall, 8.230000s user + 0.170000s system = 8.400000s CPU (3691.7%)\n",
      "Uniq kmers merging in  0.055856s wall, 0.110000s user + 0.080000s system = 0.190000s CPU (340.2%)\n",
      "Kmers branching in  0.055573s wall, 2.140000s user + 0.000000s system = 2.140000s CPU (3850.8%)\n",
      "Kmer: 51 Graph size: 514372 Contigs in: 342\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.079874s wall, 2.890000s user + 0.000000s system = 2.890000s CPU (3618.2%)\n",
      "Kmers in multiple/single contigs: 0 340226\n",
      "Fragments before: 756 53252\n",
      "Fragments after: 697 53252\n",
      "New seeds: 19\n",
      "New seeds in  0.183345s wall, 7.990000s user + 0.000000s system = 7.990000s CPU (4357.9%)\n",
      "Fragments before: 720 35276\n",
      "Fragments after: 581 35276\n",
      "Connectors: 139 Extenders: 442\n",
      "Connections and extensions in  0.188462s wall, 5.050000s user + 0.030000s system = 5.080000s CPU (2695.5%)\n",
      "Contigs out: 222 Genome: 4814548 N50: 66983 L50: 25\n",
      "Assembled in  0.458711s wall, 15.940000s user + 0.030000s system = 15.970000s CPU (3481.5%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 70428\n",
      "Reads for connection: 44548\n",
      "Internal reads: 4457\n",
      "Reads cleaned in  0.468546s wall, 10.690000s user + 0.100000s system = 10.790000s CPU (2302.9%)\n",
      "\n",
      "Kmer len: 67\n",
      "Raw kmers: 10963309 Memory needed (GB): 0.420991 Memory available (GB): 29.8166 1 cycle(s) will be performed\n",
      "Distinct kmers: 329566\n",
      "Kmer count in  0.163234s wall, 6.090000s user + 0.040000s system = 6.130000s CPU (3755.3%)\n",
      "Uniq kmers merging in  0.046565s wall, 0.090000s user + 0.050000s system = 0.140000s CPU (300.7%)\n",
      "Kmers branching in  0.043704s wall, 1.660000s user + 0.010000s system = 1.670000s CPU (3821.1%)\n",
      "Kmer: 67 Graph size: 329566 Contigs in: 222\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.105896s wall, 2.870000s user + 0.000000s system = 2.870000s CPU (2710.2%)\n",
      "Kmers in multiple/single contigs: 0 201838\n",
      "Fragments before: 513 39493\n",
      "Fragments after: 481 39493\n",
      "New seeds: 5\n",
      "New seeds in  0.126491s wall, 5.530000s user + 0.010000s system = 5.540000s CPU (4379.7%)\n",
      "Fragments before: 452 27425\n",
      "Fragments after: 410 27425\n",
      "Connectors: 42 Extenders: 368\n",
      "Connections and extensions in  0.227548s wall, 5.130000s user + 0.020000s system = 5.150000s CPU (2263.3%)\n",
      "Contigs out: 185 Genome: 4820584 N50: 78027 L50: 21\n",
      "Assembled in  0.464277s wall, 13.540000s user + 0.030000s system = 13.570000s CPU (2922.8%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 57164\n",
      "Reads for connection: 37310\n",
      "Internal reads: 3939\n",
      "Reads cleaned in  0.487847s wall, 10.200000s user + 0.110000s system = 10.310000s CPU (2113.4%)\n",
      "\n",
      "Kmer len: 81\n",
      "Raw kmers: 8614301 Memory needed (GB): 0.330789 Memory available (GB): 29.8041 1 cycle(s) will be performed\n",
      "Distinct kmers: 270496\n",
      "Kmer count in  0.158958s wall, 5.640000s user + 0.060000s system = 5.700000s CPU (3585.9%)\n",
      "Uniq kmers merging in  0.047572s wall, 0.070000s user + 0.060000s system = 0.130000s CPU (273.3%)\n",
      "Kmers branching in  0.036924s wall, 1.410000s user + 0.010000s system = 1.420000s CPU (3845.8%)\n",
      "Kmer: 81 Graph size: 270496 Contigs in: 185\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.091663s wall, 2.780000s user + 0.000000s system = 2.780000s CPU (3032.8%)\n",
      "Kmers in multiple/single contigs: 0 160387\n",
      "Fragments before: 412 32198\n",
      "Fragments after: 372 32198\n",
      "New seeds: 2\n",
      "New seeds in  0.102307s wall, 4.280000s user + 0.010000s system = 4.290000s CPU (4193.3%)\n",
      "Fragments before: 358 25436\n",
      "Fragments after: 327 25436\n",
      "Connectors: 31 Extenders: 296\n",
      "Connections and extensions in  0.231043s wall, 4.570000s user + 0.020000s system = 4.590000s CPU (1986.6%)\n",
      "Contigs out: 149 Genome: 4824918 N50: 100845 L50: 17\n",
      "Assembled in  0.428643s wall, 11.630000s user + 0.030000s system = 11.660000s CPU (2720.2%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 48264\n",
      "Reads for connection: 33700\n",
      "Internal reads: 3321\n",
      "Reads cleaned in  0.562299s wall, 9.880000s user + 0.130000s system = 10.010000s CPU (1780.2%)\n",
      "\n",
      "Kmer len: 97\n",
      "Raw kmers: 6753463 Memory needed (GB): 0.324166 Memory available (GB): 29.7915 1 cycle(s) will be performed\n",
      "Distinct kmers: 222243\n",
      "Kmer count in  0.134024s wall, 4.610000s user + 0.040000s system = 4.650000s CPU (3469.5%)\n",
      "Uniq kmers merging in  0.046364s wall, 0.090000s user + 0.040000s system = 0.130000s CPU (280.4%)\n",
      "Kmers branching in  0.036665s wall, 1.370000s user + 0.000000s system = 1.370000s CPU (3736.5%)\n",
      "Kmer: 97 Graph size: 222243 Contigs in: 149\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.121189s wall, 3.000000s user + 0.010000s system = 3.010000s CPU (2483.7%)\n",
      "Kmers in multiple/single contigs: 0 127122\n",
      "Fragments before: 320 28406\n",
      "Fragments after: 286 28406\n",
      "New seeds: 2\n",
      "New seeds in  0.094286s wall, 3.750000s user + 0.010000s system = 3.760000s CPU (3987.9%)\n",
      "Fragments before: 284 23931\n",
      "Fragments after: 260 23931\n",
      "Connectors: 24 Extenders: 236\n",
      "Connections and extensions in  0.398451s wall, 4.770000s user + 0.030000s system = 4.800000s CPU (1204.7%)\n",
      "Contigs out: 119 Genome: 4829069 N50: 132289 L50: 13\n",
      "Assembled in  0.617480s wall, 11.520000s user + 0.050000s system = 11.570000s CPU (1873.7%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 41477\n",
      "Reads for connection: 31082\n",
      "Internal reads: 2868\n",
      "Reads cleaned in  0.741508s wall, 9.510000s user + 0.180000s system = 9.690000s CPU (1306.8%)\n",
      "\n",
      "Kmer len: 111\n",
      "Raw kmers: 5486019 Memory needed (GB): 0.263329 Memory available (GB): 29.7782 1 cycle(s) will be performed\n",
      "Distinct kmers: 184837\n",
      "Kmer count in  0.139721s wall, 4.330000s user + 0.060000s system = 4.390000s CPU (3142.0%)\n",
      "Uniq kmers merging in  0.064853s wall, 0.060000s user + 0.080000s system = 0.140000s CPU (215.9%)\n",
      "Kmers branching in  0.032712s wall, 1.220000s user + 0.000000s system = 1.220000s CPU (3729.5%)\n",
      "Kmer: 111 Graph size: 184837 Contigs in: 119\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.199418s wall, 2.650000s user + 0.030000s system = 2.680000s CPU (1343.9%)\n",
      "Kmers in multiple/single contigs: 0 101546\n",
      "Fragments before: 278 24653\n",
      "Fragments after: 245 24653\n",
      "New seeds: 1\n",
      "New seeds in  0.082747s wall, 3.250000s user + 0.000000s system = 3.250000s CPU (3927.6%)\n",
      "Fragments before: 233 20168\n",
      "Fragments after: 216 20168\n",
      "Connectors: 17 Extenders: 199\n",
      "Connections and extensions in  0.402541s wall, 4.500000s user + 0.050000s system = 4.550000s CPU (1130.3%)\n",
      "Contigs out: 101 Genome: 4832150 N50: 149715 L50: 12\n",
      "Assembled in  0.687723s wall, 10.410000s user + 0.080000s system = 10.490000s CPU (1525.3%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 36740\n",
      "Reads for connection: 29274\n",
      "Internal reads: 2514\n",
      "Reads cleaned in  0.787928s wall, 9.010000s user + 0.160000s system = 9.170000s CPU (1163.8%)\n",
      "\n",
      "Kmer len: 127\n",
      "Raw kmers: 4492990 Memory needed (GB): 0.215664 Memory available (GB): 29.7664 1 cycle(s) will be performed\n",
      "Distinct kmers: 159227\n",
      "Kmer count in  0.153401s wall, 3.610000s user + 0.100000s system = 3.710000s CPU (2418.5%)\n",
      "Uniq kmers merging in  0.045501s wall, 0.060000s user + 0.040000s system = 0.100000s CPU (219.8%)\n",
      "Kmers branching in  0.027287s wall, 1.080000s user + 0.010000s system = 1.090000s CPU (3994.6%)\n",
      "Kmer: 127 Graph size: 159227 Contigs in: 101\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.182510s wall, 2.480000s user + 0.010000s system = 2.490000s CPU (1364.3%)\n",
      "Kmers in multiple/single contigs: 0 84892\n",
      "Fragments before: 229 21615\n",
      "Fragments after: 199 21615\n",
      "New seeds: 0\n",
      "New seeds in  0.069565s wall, 2.930000s user + 0.010000s system = 2.940000s CPU (4226.3%)\n",
      "Fragments before: 191 19078\n",
      "Fragments after: 180 19078\n",
      "Connectors: 11 Extenders: 169\n",
      "Connections and extensions in  0.469573s wall, 4.270000s user + 0.050000s system = 4.320000s CPU (920.0%)\n",
      "Contigs out: 87 Genome: 4834786 N50: 160303 L50: 10\n",
      "Assembled in  0.724242s wall, 9.680000s user + 0.070000s system = 9.750000s CPU (1346.2%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 32363\n",
      "Reads for connection: 27660\n",
      "Internal reads: 2233\n",
      "Reads cleaned in  0.904008s wall, 8.400000s user + 0.230000s system = 8.630000s CPU (954.6%)\n",
      "\n",
      "Kmer len: 143\n",
      "Raw kmers: 3670932 Memory needed (GB): 0.211446 Memory available (GB): 29.7554 1 cycle(s) will be performed\n",
      "Distinct kmers: 138396\n",
      "Kmer count in  0.142660s wall, 3.350000s user + 0.100000s system = 3.450000s CPU (2418.3%)\n",
      "Uniq kmers merging in  0.034369s wall, 0.050000s user + 0.040000s system = 0.090000s CPU (261.9%)\n",
      "Kmers branching in  0.030855s wall, 0.980000s user + 0.000000s system = 0.980000s CPU (3176.1%)\n",
      "Kmer: 143 Graph size: 138396 Contigs in: 87\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.217405s wall, 2.570000s user + 0.020000s system = 2.590000s CPU (1191.3%)\n",
      "Kmers in multiple/single contigs: 0 72856\n",
      "Fragments before: 202 19919\n",
      "Fragments after: 172 19919\n",
      "New seeds: 0\n",
      "New seeds in  0.060213s wall, 2.480000s user + 0.010000s system = 2.490000s CPU (4135.3%)\n",
      "Fragments before: 162 16610\n",
      "Fragments after: 153 16610\n",
      "Connectors: 9 Extenders: 144\n",
      "Connections and extensions in  0.528436s wall, 4.460000s user + 0.050000s system = 4.510000s CPU (853.5%)\n",
      "Contigs out: 75 Genome: 4836364 N50: 161583 L50: 9\n",
      "Assembled in  0.808107s wall, 9.510000s user + 0.080000s system = 9.590000s CPU (1186.7%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 27994\n",
      "Reads for connection: 26352\n",
      "Internal reads: 1868\n",
      "Reads cleaned in  0.937104s wall, 8.300000s user + 0.220000s system = 8.520000s CPU (909.2%)\n",
      "\n",
      "Kmer len: 157\n",
      "Raw kmers: 2971015 Memory needed (GB): 0.17113 Memory available (GB): 29.7441 1 cycle(s) will be performed\n",
      "Distinct kmers: 118719\n",
      "Kmer count in  0.143515s wall, 2.940000s user + 0.100000s system = 3.040000s CPU (2118.2%)\n",
      "Uniq kmers merging in  0.038271s wall, 0.060000s user + 0.030000s system = 0.090000s CPU (235.2%)\n",
      "Kmers branching in  0.027394s wall, 0.940000s user + 0.010000s system = 0.950000s CPU (3468.0%)\n",
      "Kmer: 157 Graph size: 118719 Contigs in: 75\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.205977s wall, 2.540000s user + 0.010000s system = 2.550000s CPU (1238.0%)\n",
      "Kmers in multiple/single contigs: 0 60930\n",
      "Fragments before: 198 18671\n",
      "Fragments after: 160 18671\n",
      "New seeds: 0\n",
      "New seeds in  0.057364s wall, 2.280000s user + 0.010000s system = 2.290000s CPU (3992.1%)\n",
      "Fragments before: 141 14886\n",
      "Fragments after: 137 14886\n",
      "Connectors: 4 Extenders: 133\n",
      "Connections and extensions in  0.485618s wall, 4.350000s user + 0.030000s system = 4.380000s CPU (901.9%)\n",
      "Contigs out: 70 Genome: 4837350 N50: 197750 L50: 9\n",
      "Assembled in  0.750762s wall, 9.170000s user + 0.050000s system = 9.220000s CPU (1228.1%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 25999\n",
      "Reads for connection: 25942\n",
      "Internal reads: 1732\n",
      "Reads cleaned in  0.945731s wall, 7.920000s user + 0.260000s system = 8.180000s CPU (864.9%)\n",
      "\n",
      "Kmer len: 173\n",
      "Raw kmers: 2489781 Memory needed (GB): 0.167313 Memory available (GB): 29.7338 1 cycle(s) will be performed\n",
      "Distinct kmers: 107041\n",
      "Kmer count in  0.143501s wall, 2.540000s user + 0.130000s system = 2.670000s CPU (1860.6%)\n",
      "Uniq kmers merging in  0.037256s wall, 0.040000s user + 0.050000s system = 0.090000s CPU (241.6%)\n",
      "Kmers branching in  0.027126s wall, 0.810000s user + 0.000000s system = 0.810000s CPU (2986.1%)\n",
      "Kmer: 173 Graph size: 107041 Contigs in: 70\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.217547s wall, 2.520000s user + 0.000000s system = 2.520000s CPU (1158.4%)\n",
      "Kmers in multiple/single contigs: 0 53806\n",
      "Fragments before: 201 19663\n",
      "Fragments after: 159 19663\n",
      "New seeds: 0\n",
      "New seeds in  0.057957s wall, 2.210000s user + 0.000000s system = 2.210000s CPU (3813.2%)\n",
      "Fragments before: 127 14331\n",
      "Fragments after: 127 14331\n",
      "Connectors: 0 Extenders: 127\n",
      "Connections and extensions in  0.532087s wall, 4.350000s user + 0.050000s system = 4.400000s CPU (826.9%)\n",
      "Contigs out: 67 Genome: 4837258 N50: 197750 L50: 9\n",
      "Assembled in  0.809416s wall, 9.090000s user + 0.050000s system = 9.140000s CPU (1129.2%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 24440\n",
      "Reads for connection: 25864\n",
      "Internal reads: 1617\n",
      "Reads cleaned in  0.930341s wall, 8.040000s user + 0.270000s system = 8.310000s CPU (893.2%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 21\n",
      "Connected: 9635 ambiguously connected: 2548 from 12932 mate pairs\n",
      "Connect pairs in  1.855532s wall, 91.470000s user + 0.100000s system = 91.570000s CPU (4935.0%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 35\n",
      "Connected: 471 ambiguously connected: 2053 from 2548 mate pairs\n",
      "Connect pairs in  0.493353s wall, 18.100000s user + 0.020000s system = 18.120000s CPU (3672.8%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 51\n",
      "Connected: 194 ambiguously connected: 1831 from 2053 mate pairs\n",
      "Connect pairs in  0.316971s wall, 11.920000s user + 0.120000s system = 12.040000s CPU (3798.5%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 67\n",
      "Connected: 203 ambiguously connected: 1584 from 1831 mate pairs\n",
      "Connect pairs in  0.315218s wall, 12.320000s user + 0.020000s system = 12.340000s CPU (3914.8%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 81\n",
      "Connected: 108 ambiguously connected: 1435 from 1584 mate pairs\n",
      "Connect pairs in  0.258257s wall, 10.490000s user + 0.020000s system = 10.510000s CPU (4069.6%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 97\n",
      "Connected: 134 ambiguously connected: 1269 from 1435 mate pairs\n",
      "Connect pairs in  0.307715s wall, 10.680000s user + 0.030000s system = 10.710000s CPU (3480.5%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 111\n",
      "Connected: 57 ambiguously connected: 1188 from 1269 mate pairs\n",
      "Connect pairs in  0.258485s wall, 9.850000s user + 0.020000s system = 9.870000s CPU (3818.4%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 127\n",
      "Connected: 159 ambiguously connected: 973 from 1188 mate pairs\n",
      "Connect pairs in  0.254978s wall, 9.610000s user + 0.030000s system = 9.640000s CPU (3780.7%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 143\n",
      "Connected: 112 ambiguously connected: 837 from 973 mate pairs\n",
      "Connect pairs in  0.243890s wall, 8.460000s user + 0.040000s system = 8.500000s CPU (3485.2%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 157\n",
      "Connected: 63 ambiguously connected: 741 from 837 mate pairs\n",
      "Connect pairs in  0.206419s wall, 7.270000s user + 0.030000s system = 7.300000s CPU (3536.5%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 173\n",
      "Connected: 57 ambiguously connected: 648 from 741 mate pairs\n",
      "Connect pairs in  0.223372s wall, 7.480000s user + 0.000000s system = 7.480000s CPU (3348.7%)\n",
      "Totally connected: 11193\n",
      "Added notconnected: 1295\n",
      "\n",
      "Kmer len: 215\n",
      "Raw kmers: 2409066 Memory needed (GB): 0.185016 Memory available (GB): 29.7224 1 cycle(s) will be performed\n",
      "Distinct kmers: 91339\n",
      "Kmer count in  0.117852s wall, 2.550000s user + 0.540000s system = 3.090000s CPU (2621.9%)\n",
      "Uniq kmers merging in  0.049258s wall, 0.030000s user + 0.070000s system = 0.100000s CPU (203.0%)\n",
      "Kmers branching in  0.022380s wall, 0.780000s user + 0.010000s system = 0.790000s CPU (3530.0%)\n",
      "Kmer: 215 Graph size: 91339 Contigs in: 67\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.210844s wall, 2.450000s user + 0.000000s system = 2.450000s CPU (1162.0%)\n",
      "Kmers in multiple/single contigs: 0 55228\n",
      "Fragments before: 205 25985\n",
      "Fragments after: 145 25985\n",
      "New seeds: 0\n",
      "New seeds in  0.062893s wall, 1.700000s user + 0.010000s system = 1.710000s CPU (2718.9%)\n",
      "Fragments before: 123 19306\n",
      "Fragments after: 114 19306\n",
      "Connectors: 9 Extenders: 105\n",
      "Connections and extensions in  0.539311s wall, 4.350000s user + 0.430000s system = 4.780000s CPU (886.3%)\n",
      "Contigs out: 56 Genome: 4840295 N50: 231146 L50: 7\n",
      "Assembled in  0.814382s wall, 8.500000s user + 0.440000s system = 8.940000s CPU (1097.8%)\n",
      "\n",
      "\n",
      "Kmer len: 289\n",
      "Raw kmers: 1783850 Memory needed (GB): 0.188375 Memory available (GB): 29.7116 1 cycle(s) will be performed\n",
      "Distinct kmers: 90590\n",
      "Kmer count in  0.082690s wall, 2.210000s user + 0.120000s system = 2.330000s CPU (2817.7%)\n",
      "Uniq kmers merging in  0.049282s wall, 0.050000s user + 0.060000s system = 0.110000s CPU (223.2%)\n",
      "Kmers branching in  0.025774s wall, 0.890000s user + 0.010000s system = 0.900000s CPU (3491.9%)\n",
      "Kmer: 289 Graph size: 90590 Contigs in: 56\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.267261s wall, 2.580000s user + 0.000000s system = 2.580000s CPU (965.3%)\n",
      "Kmers in multiple/single contigs: 0 51840\n",
      "Fragments before: 176 27790\n",
      "Fragments after: 118 27790\n",
      "New seeds: 0\n",
      "New seeds in  0.059435s wall, 1.600000s user + 0.010000s system = 1.610000s CPU (2708.9%)\n",
      "Fragments before: 100 19281\n",
      "Fragments after: 91 19281\n",
      "Connectors: 9 Extenders: 82\n",
      "Connections and extensions in  0.665138s wall, 4.180000s user + 0.360000s system = 4.540000s CPU (682.6%)\n",
      "Contigs out: 45 Genome: 4843194 N50: 298365 L50: 5\n",
      "Assembled in  0.993256s wall, 8.360000s user + 0.370000s system = 8.730000s CPU (878.9%)\n",
      "\n",
      "\n",
      "Kmer len: 363\n",
      "Raw kmers: 1304455 Memory needed (GB): 0.162796 Memory available (GB): 29.6987 1 cycle(s) will be performed\n",
      "Distinct kmers: 89270\n",
      "Kmer count in  0.069587s wall, 1.770000s user + 0.060000s system = 1.830000s CPU (2629.8%)\n",
      "Uniq kmers merging in  0.058736s wall, 0.050000s user + 0.070000s system = 0.120000s CPU (204.3%)\n",
      "Kmers branching in  0.025795s wall, 1.040000s user + 0.010000s system = 1.050000s CPU (4070.5%)\n",
      "Kmer: 363 Graph size: 89270 Contigs in: 45\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.324727s wall, 2.560000s user + 0.000000s system = 2.560000s CPU (788.4%)\n",
      "Kmers in multiple/single contigs: 0 50178\n",
      "Fragments before: 168 27810\n",
      "Fragments after: 103 27810\n",
      "New seeds: 0\n",
      "New seeds in  0.064799s wall, 1.900000s user + 0.010000s system = 1.910000s CPU (2947.6%)\n",
      "Fragments before: 72 16045\n",
      "Fragments after: 70 16045\n",
      "Connectors: 2 Extenders: 68\n",
      "Connections and extensions in  0.884531s wall, 3.890000s user + 0.190000s system = 4.080000s CPU (461.3%)\n",
      "Contigs out: 38 Genome: 4842245 N50: 298538 L50: 4\n",
      "Assembled in  1.275517s wall, 8.350000s user + 0.200000s system = 8.550000s CPU (670.3%)\n",
      "\n",
      "DONE\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda run -n skesa skesa --reads SRR3028792_1.fp.fastq.gz,SRR3028792_2.fp.fastq.gz --contigs_out SRR3028792.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a15ee6b-9b19-47ec-a2a7-d9825edafaa9",
   "metadata": {},
   "source": [
    "# 5. Evaulate assembly quality (`quast`)\n",
    "\n",
    "## 5.1. Install `quast`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "603778c6-3265-432a-a70b-f1740251e69d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/quast\n",
      "\n",
      "  added / updated specs:\n",
      "    - quast\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
      "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu\n",
      "  blast              bioconda/linux-64::blast-2.12.0-pl5262h3289130_0\n",
      "  brotli             conda-forge/linux-64::brotli-1.0.9-h7f98852_6\n",
      "  brotli-bin         conda-forge/linux-64::brotli-bin-1.0.9-h7f98852_6\n",
      "  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4\n",
      "  c-ares             conda-forge/linux-64::c-ares-1.18.1-h7f98852_0\n",
      "  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0\n",
      "  certifi            conda-forge/linux-64::certifi-2021.10.8-py37h89c1867_1\n",
      "  circos             bioconda/noarch::circos-0.69.8-hdfd78af_1\n",
      "  curl               conda-forge/linux-64::curl-7.81.0-h494985f_0\n",
      "  cycler             conda-forge/noarch::cycler-0.11.0-pyhd8ed1ab_0\n",
      "  entrez-direct      bioconda/linux-64::entrez-direct-16.2-he881be0_0\n",
      "  expat              conda-forge/linux-64::expat-2.4.3-h9c3ff4c_0\n",
      "  fontconfig         conda-forge/linux-64::fontconfig-2.13.1-hba837de_1005\n",
      "  fonttools          conda-forge/linux-64::fonttools-4.28.5-py37h5e8e339_0\n",
      "  freetype           conda-forge/linux-64::freetype-2.10.4-h0708190_1\n",
      "  giflib             conda-forge/linux-64::giflib-5.2.1-h36c2ea0_2\n",
      "  glimmerhmm         bioconda/linux-64::glimmerhmm-3.0.4-pl526hf484d3e_3\n",
      "  icu                conda-forge/linux-64::icu-67.1-he1b5a44_0\n",
      "  jbig               conda-forge/linux-64::jbig-2.1-h7f98852_2003\n",
      "  joblib             conda-forge/noarch::joblib-1.1.0-pyhd8ed1ab_0\n",
      "  jpeg               conda-forge/linux-64::jpeg-9d-h36c2ea0_0\n",
      "  k8                 bioconda/linux-64::k8-0.2.5-h9a82719_1\n",
      "  kiwisolver         conda-forge/linux-64::kiwisolver-1.3.2-py37h2527ec5_1\n",
      "  krb5               conda-forge/linux-64::krb5-1.19.2-h48eae69_3\n",
      "  lcms2              conda-forge/linux-64::lcms2-2.12-hddcbb42_0\n",
      "  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.36.1-hea4e1c9_2\n",
      "  lerc               conda-forge/linux-64::lerc-3.0-h9c3ff4c_0\n",
      "  libblas            conda-forge/linux-64::libblas-3.9.0-13_linux64_openblas\n",
      "  libbrotlicommon    conda-forge/linux-64::libbrotlicommon-1.0.9-h7f98852_6\n",
      "  libbrotlidec       conda-forge/linux-64::libbrotlidec-1.0.9-h7f98852_6\n",
      "  libbrotlienc       conda-forge/linux-64::libbrotlienc-1.0.9-h7f98852_6\n",
      "  libcblas           conda-forge/linux-64::libcblas-3.9.0-13_linux64_openblas\n",
      "  libcurl            conda-forge/linux-64::libcurl-7.81.0-h494985f_0\n",
      "  libdeflate         conda-forge/linux-64::libdeflate-1.8-h7f98852_0\n",
      "  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2\n",
      "  libev              conda-forge/linux-64::libev-4.33-h516909a_1\n",
      "  libffi             conda-forge/linux-64::libffi-3.4.2-h7f98852_5\n",
      "  libgcc             conda-forge/linux-64::libgcc-7.2.0-h69d50b8_2\n",
      "  libgcc-ng          conda-forge/linux-64::libgcc-ng-11.2.0-h1d223b6_11\n",
      "  libgd              conda-forge/linux-64::libgd-2.2.5-h6e6070c_1008\n",
      "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-11.2.0-h69a702a_11\n",
      "  libgfortran5       conda-forge/linux-64::libgfortran5-11.2.0-h5c6108e_11\n",
      "  libgomp            conda-forge/linux-64::libgomp-11.2.0-h1d223b6_11\n",
      "  libiconv           conda-forge/linux-64::libiconv-1.16-h516909a_0\n",
      "  liblapack          conda-forge/linux-64::liblapack-3.9.0-13_linux64_openblas\n",
      "  libnghttp2         conda-forge/linux-64::libnghttp2-1.43.0-ha19adfc_1\n",
      "  libnsl             conda-forge/linux-64::libnsl-2.0.0-h7f98852_0\n",
      "  libopenblas        conda-forge/linux-64::libopenblas-0.3.18-pthreads_h8fe5266_0\n",
      "  libpng             conda-forge/linux-64::libpng-1.6.37-h21135ba_2\n",
      "  libssh2            conda-forge/linux-64::libssh2-1.10.0-ha35d2d1_2\n",
      "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-11.2.0-he4da1e4_11\n",
      "  libtiff            conda-forge/linux-64::libtiff-4.3.0-h6f004c6_2\n",
      "  libuuid            conda-forge/linux-64::libuuid-2.32.1-h7f98852_1000\n",
      "  libwebp            conda-forge/linux-64::libwebp-1.2.1-h3452ae3_0\n",
      "  libwebp-base       conda-forge/linux-64::libwebp-base-1.2.1-h7f98852_0\n",
      "  libxml2            conda-forge/linux-64::libxml2-2.9.10-h68273f3_2\n",
      "  libzlib            conda-forge/linux-64::libzlib-1.2.11-h36c2ea0_1013\n",
      "  lz4-c              conda-forge/linux-64::lz4-c-1.9.3-h9c3ff4c_1\n",
      "  matplotlib-base    conda-forge/linux-64::matplotlib-base-3.5.1-py37h1058ff1_0\n",
      "  minimap2           bioconda/linux-64::minimap2-2.24-h5bf99c6_0\n",
      "  munkres            bioconda/noarch::munkres-1.0.7-py_1\n",
      "  ncurses            conda-forge/linux-64::ncurses-6.3-h9c3ff4c_0\n",
      "  numpy              conda-forge/linux-64::numpy-1.21.5-py37hf2998dd_0\n",
      "  olefile            conda-forge/noarch::olefile-0.46-pyh9f0ad1d_1\n",
      "  openjdk            conda-forge/linux-64::openjdk-10.0.2-h14c3975_1015\n",
      "  openjpeg           conda-forge/linux-64::openjpeg-2.4.0-hb52868f_1\n",
      "  openssl            conda-forge/linux-64::openssl-3.0.0-h7f98852_2\n",
      "  packaging          conda-forge/noarch::packaging-21.3-pyhd8ed1ab_0\n",
      "  pcre               conda-forge/linux-64::pcre-8.45-h9c3ff4c_0\n",
      "  perl               conda-forge/linux-64::perl-5.26.2-h36c2ea0_1008\n",
      "  perl-archive-tar   bioconda/linux-64::perl-archive-tar-2.32-pl526_0\n",
      "  perl-autoloader    bioconda/linux-64::perl-autoloader-5.74-pl526_2\n",
      "  perl-carp          bioconda/linux-64::perl-carp-1.38-pl526_3\n",
      "  perl-clone         bioconda/linux-64::perl-clone-0.42-pl526h516909a_0\n",
      "  perl-common-sense  bioconda/linux-64::perl-common-sense-3.74-pl526_2\n",
      "  perl-compress-raw~ bioconda/linux-64::perl-compress-raw-bzip2-2.087-pl526he1b5a44_0\n",
      "  perl-compress-raw~ bioconda/linux-64::perl-compress-raw-zlib-2.087-pl526hc9558a2_0\n",
      "  perl-config-gener~ bioconda/linux-64::perl-config-general-2.63-pl526_0\n",
      "  perl-digest-perl-~ bioconda/linux-64::perl-digest-perl-md5-1.9-pl526_1\n",
      "  perl-dynaloader    bioconda/linux-64::perl-dynaloader-1.25-pl526_1\n",
      "  perl-exporter      bioconda/linux-64::perl-exporter-5.72-pl526_1\n",
      "  perl-exporter-tiny bioconda/linux-64::perl-exporter-tiny-1.002001-pl526_0\n",
      "  perl-extutils-mak~ bioconda/linux-64::perl-extutils-makemaker-7.36-pl526_1\n",
      "  perl-font-ttf      bioconda/linux-64::perl-font-ttf-1.06-pl526_0\n",
      "  perl-gd            bioconda/linux-64::perl-gd-2.71-pl526he860b03_0\n",
      "  perl-io-compress   bioconda/linux-64::perl-io-compress-2.087-pl526he1b5a44_0\n",
      "  perl-io-string     bioconda/linux-64::perl-io-string-1.08-pl526_3\n",
      "  perl-io-zlib       bioconda/linux-64::perl-io-zlib-1.10-pl526_2\n",
      "  perl-json          bioconda/linux-64::perl-json-4.02-pl526_0\n",
      "  perl-json-xs       bioconda/linux-64::perl-json-xs-2.34-pl526h6bb024c_3\n",
      "  perl-list-moreuti~ bioconda/linux-64::perl-list-moreutils-0.428-pl526_1\n",
      "  perl-list-moreuti~ bioconda/linux-64::perl-list-moreutils-xs-0.428-pl526_0\n",
      "  perl-math-bezier   bioconda/linux-64::perl-math-bezier-0.01-pl526_1\n",
      "  perl-math-round    bioconda/linux-64::perl-math-round-0.07-pl526_1\n",
      "  perl-math-vecstat  bioconda/linux-64::perl-math-vecstat-0.08-pl526_1\n",
      "  perl-module-imple~ bioconda/linux-64::perl-module-implementation-0.09-pl526_2\n",
      "  perl-module-runti~ bioconda/linux-64::perl-module-runtime-0.016-pl526_1\n",
      "  perl-number-format bioconda/linux-64::perl-number-format-1.75-pl526_3\n",
      "  perl-params-valid~ bioconda/linux-64::perl-params-validate-1.29-pl526h14c3975_1\n",
      "  perl-pathtools     bioconda/linux-64::perl-pathtools-3.75-pl526h14c3975_1\n",
      "  perl-readonly      bioconda/linux-64::perl-readonly-2.05-pl526_0\n",
      "  perl-regexp-common bioconda/linux-64::perl-regexp-common-2017060201-pl526_0\n",
      "  perl-scalar-list-~ bioconda/linux-64::perl-scalar-list-utils-1.52-pl526h516909a_0\n",
      "  perl-set-intspan   bioconda/linux-64::perl-set-intspan-1.19-pl526_1\n",
      "  perl-statistics-b~ bioconda/linux-64::perl-statistics-basic-1.6611-pl526_2\n",
      "  perl-svg           bioconda/linux-64::perl-svg-2.84-pl526_0\n",
      "  perl-text-format   bioconda/linux-64::perl-text-format-0.59-pl526_2\n",
      "  perl-threaded      bioconda/noarch::perl-threaded-5.32.1-hdfd78af_1\n",
      "  perl-time-hires    bioconda/linux-64::perl-time-hires-1.9760-pl526h14c3975_1\n",
      "  perl-try-tiny      bioconda/linux-64::perl-try-tiny-0.30-pl526_1\n",
      "  perl-types-serial~ bioconda/linux-64::perl-types-serialiser-1.0-pl526_2\n",
      "  perl-xml-parser    bioconda/linux-64::perl-xml-parser-2.44-3\n",
      "  perl-xsloader      bioconda/linux-64::perl-xsloader-0.24-pl526_0\n",
      "  pillow             conda-forge/linux-64::pillow-8.4.0-py37h0f21c89_0\n",
      "  pip                conda-forge/noarch::pip-21.3.1-pyhd8ed1ab_0\n",
      "  pyparsing          conda-forge/noarch::pyparsing-3.0.6-pyhd8ed1ab_0\n",
      "  python             conda-forge/linux-64::python-3.7.12-hf930737_100_cpython\n",
      "  python-dateutil    conda-forge/noarch::python-dateutil-2.8.2-pyhd8ed1ab_0\n",
      "  python_abi         conda-forge/linux-64::python_abi-3.7-2_cp37m\n",
      "  quast              bioconda/linux-64::quast-5.0.2-py37pl5262hfecc14a_5\n",
      "  readline           conda-forge/linux-64::readline-8.1-h46c0cb4_0\n",
      "  setuptools         conda-forge/linux-64::setuptools-60.5.0-py37h89c1867_0\n",
      "  simplejson         conda-forge/linux-64::simplejson-3.17.6-py37h5e8e339_0\n",
      "  six                conda-forge/noarch::six-1.16.0-pyh6c4a22f_0\n",
      "  sqlite             conda-forge/linux-64::sqlite-3.37.0-h9cd32fc_0\n",
      "  tk                 conda-forge/linux-64::tk-8.6.11-h27826a3_1\n",
      "  unicodedata2       conda-forge/linux-64::unicodedata2-14.0.0-py37h5e8e339_0\n",
      "  wheel              conda-forge/noarch::wheel-0.37.1-pyhd8ed1ab_0\n",
      "  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1\n",
      "  zlib               conda-forge/linux-64::zlib-1.2.11-h36c2ea0_1013\n",
      "  zstd               conda-forge/linux-64::zstd-1.5.1-ha95c52a_0\n",
      "\n",
      "\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: \\ The default QUAST package does not include:\n",
      "* GRIDSS (needed for structural variants detection)\n",
      "* SILVA 16S rRNA database (needed for reference genome detection in metagenomic datasets)\n",
      "* BUSCO tools and databases (needed for searching BUSCO genes) -- works in Linux only!\n",
      "\n",
      "To be able to use those, please run\n",
      "    quast-download-gridss\n",
      "    quast-download-silva\n",
      "    quast-download-busco\n",
      "\n",
      "done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate quast\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda create -y -n quast quast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "5640c41e-c3f0-4d64-9db1-2cb35adea2f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QUAST: Quality Assessment Tool for Genome Assemblies\n",
      "Version: 5.0.2\n",
      "\n",
      "Usage: python /home/CSCScience.ca/apetkau/miniconda3/envs/quast/bin/quast [options] <files_with_contigs>\n",
      "\n",
      "Options:\n",
      "-o  --output-dir  <dirname>       Directory to store all result files [default: quast_results/results_<datetime>]\n",
      "-r                <filename>      Reference genome file\n",
      "-g  --features [type:]<filename>  File with genomic feature coordinates in the reference (GFF, BED, NCBI or TXT)\n",
      "                                  Optional 'type' can be specified for extracting only a specific feature type from GFF\n",
      "-m  --min-contig  <int>           Lower threshold for contig length [default: 500]\n",
      "-t  --threads     <int>           Maximum number of threads [default: 25% of CPUs]\n",
      "\n",
      "Advanced options:\n",
      "-s  --split-scaffolds                 Split assemblies by continuous fragments of N's and add such \"contigs\" to the comparison\n",
      "-l  --labels \"label, label, ...\"      Names of assemblies to use in reports, comma-separated. If contain spaces, use quotes\n",
      "-L                                    Take assembly names from their parent directory names\n",
      "-e  --eukaryote                       Genome is eukaryotic (primarily affects gene prediction)\n",
      "    --fungus                          Genome is fungal (primarily affects gene prediction)\n",
      "    --large                           Use optimal parameters for evaluation of large genomes\n",
      "                                      In particular, imposes '-e -m 3000 -i 500 -x 7000' (can be overridden manually)\n",
      "-k  --k-mer-stats                     Compute k-mer-based quality metrics (recommended for large genomes)\n",
      "                                      This may significantly increase memory and time consumption on large genomes\n",
      "    --k-mer-size                      Size of k used in --k-mer-stats [default: 101]\n",
      "    --circos                          Draw Circos plot\n",
      "-f  --gene-finding                    Predict genes using GeneMarkS (prokaryotes, default) or GeneMark-ES (eukaryotes, use --eukaryote)\n",
      "    --mgm                             Use MetaGeneMark for gene prediction (instead of the default finder, see above)\n",
      "    --glimmer                         Use GlimmerHMM for gene prediction (instead of the default finder, see above)\n",
      "    --gene-thresholds <int,int,...>   Comma-separated list of threshold lengths of genes to search with Gene Finding module\n",
      "                                      [default: 0,300,1500,3000]\n",
      "    --rna-finding                     Predict ribosomal RNA genes using Barrnap\n",
      "-b  --conserved-genes-finding         Count conserved orthologs using BUSCO (only on Linux)\n",
      "    --operons  <filename>             File with operon coordinates in the reference (GFF, BED, NCBI or TXT)\n",
      "    --est-ref-size <int>              Estimated reference size (for computing NGx metrics without a reference)\n",
      "    --contig-thresholds <int,int,...> Comma-separated list of contig length thresholds [default: 0,1000,5000,10000,25000,50000]\n",
      "-u  --use-all-alignments              Compute genome fraction, # genes, # operons in QUAST v1.* style.\n",
      "                                      By default, QUAST filters Minimap's alignments to keep only best ones\n",
      "-i  --min-alignment <int>             The minimum alignment length [default: 65]\n",
      "    --min-identity <float>            The minimum alignment identity (80.0, 100.0) [default: 95.0]\n",
      "-a  --ambiguity-usage <none|one|all>  Use none, one, or all alignments of a contig when all of them\n",
      "                                      are almost equally good (see --ambiguity-score) [default: one]\n",
      "    --ambiguity-score <float>         Score S for defining equally good alignments of a single contig. All alignments are sorted \n",
      "                                      by decreasing LEN * IDY% value. All alignments with LEN * IDY% < S * best(LEN * IDY%) are \n",
      "                                      discarded. S should be between 0.8 and 1.0 [default: 0.99]\n",
      "    --strict-NA                       Break contigs in any misassembly event when compute NAx and NGAx.\n",
      "                                      By default, QUAST breaks contigs only by extensive misassemblies (not local ones)\n",
      "-x  --extensive-mis-size  <int>       Lower threshold for extensive misassembly size. All relocations with inconsistency\n",
      "                                      less than extensive-mis-size are counted as local misassemblies [default: 1000]\n",
      "    --scaffold-gap-max-size  <int>    Max allowed scaffold gap length difference. All relocations with inconsistency\n",
      "                                      less than scaffold-gap-size are counted as scaffold gap misassemblies [default: 10000]\n",
      "    --unaligned-part-size  <int>      Lower threshold for detecting partially unaligned contigs. Such contig should have\n",
      "                                      at least one unaligned fragment >= the threshold [default: 500]\n",
      "    --skip-unaligned-mis-contigs      Do not distinguish contigs with >= 50% unaligned bases as a separate group\n",
      "                                      By default, QUAST does not count misassemblies in them\n",
      "    --fragmented                      Reference genome may be fragmented into small pieces (e.g. scaffolded reference) \n",
      "    --fragmented-max-indent  <int>    Mark translocation as fake if both alignments are located no further than N bases \n",
      "                                      from the ends of the reference fragments [default: 85]\n",
      "                                      Requires --fragmented option\n",
      "    --upper-bound-assembly            Simulate upper bound assembly based on the reference genome and reads\n",
      "    --upper-bound-min-con  <int>      Minimal number of 'connecting reads' needed for joining upper bound contigs into a scaffold\n",
      "                                      [default: 2 for mate-pairs and 1 for long reads]\n",
      "    --est-insert-size  <int>          Use provided insert size in upper bound assembly simulation [default: auto detect from reads or 255]\n",
      "    --plots-format  <str>             Save plots in specified format [default: pdf].\n",
      "                                      Supported formats: emf, eps, pdf, png, ps, raw, rgba, svg, svgz\n",
      "    --memory-efficient                Run everything using one thread, separately per each assembly.\n",
      "                                      This may significantly reduce memory consumption on large genomes\n",
      "    --space-efficient                 Create only reports and plots files. Aux files including .stdout, .stderr, .coords will not be created.\n",
      "                                      This may significantly reduce space consumption on large genomes. Icarus viewers also will not be built\n",
      "-1  --pe1     <filename>              File with forward paired-end reads (in FASTQ format, may be gzipped)\n",
      "-2  --pe2     <filename>              File with reverse paired-end reads (in FASTQ format, may be gzipped)\n",
      "    --pe12    <filename>              File with interlaced forward and reverse paired-end reads. (in FASTQ format, may be gzipped)\n",
      "    --mp1     <filename>              File with forward mate-pair reads (in FASTQ format, may be gzipped)\n",
      "    --mp2     <filename>              File with reverse mate-pair reads (in FASTQ format, may be gzipped)\n",
      "    --mp12    <filename>              File with interlaced forward and reverse mate-pair reads (in FASTQ format, may be gzipped)\n",
      "    --single  <filename>              File with unpaired short reads (in FASTQ format, may be gzipped)\n",
      "    --pacbio     <filename>           File with PacBio reads (in FASTQ format, may be gzipped)\n",
      "    --nanopore   <filename>           File with Oxford Nanopore reads (in FASTQ format, may be gzipped)\n",
      "    --ref-sam <filename>              SAM alignment file obtained by aligning reads to reference genome file\n",
      "    --ref-bam <filename>              BAM alignment file obtained by aligning reads to reference genome file\n",
      "    --sam     <filename,filename,...> Comma-separated list of SAM alignment files obtained by aligning reads to assemblies\n",
      "                                      (use the same order as for files with contigs)\n",
      "    --bam     <filename,filename,...> Comma-separated list of BAM alignment files obtained by aligning reads to assemblies\n",
      "                                      (use the same order as for files with contigs)\n",
      "                                      Reads (or SAM/BAM file) are used for structural variation detection and\n",
      "                                      coverage histogram building in Icarus\n",
      "    --sv-bedpe  <filename>            File with structural variations (in BEDPE format)\n",
      "\n",
      "Speedup options:\n",
      "    --no-check                        Do not check and correct input fasta files. Use at your own risk (see manual)\n",
      "    --no-plots                        Do not draw plots\n",
      "    --no-html                         Do not build html reports and Icarus viewers\n",
      "    --no-icarus                       Do not build Icarus viewers\n",
      "    --no-snps                         Do not report SNPs (may significantly reduce memory consumption on large genomes)\n",
      "    --no-gc                           Do not compute GC% and GC-distribution\n",
      "    --no-sv                           Do not run structural variation detection (make sense only if reads are specified)\n",
      "    --no-gzip                         Do not compress large output files\n",
      "    --no-read-stats                   Do not align reads to assemblies\n",
      "                                      Reads will be aligned to reference and used for coverage analysis,\n",
      "                                      upper bound assembly simulation, and structural variation detection.\n",
      "                                      Use this option if you do not need read statistics for assemblies.\n",
      "    --fast                            A combination of all speedup options except --no-check\n",
      "\n",
      "Other:\n",
      "    --silent                          Do not print detailed information about each step to stdout (log file is not affected)\n",
      "    --test                            Run QUAST on the data from the test_data folder, output to quast_test_output\n",
      "    --test-sv                         Run QUAST with structural variants detection on the data from the test_data folder,\n",
      "                                      output to quast_test_output\n",
      "-h  --help                            Print full usage message\n",
      "-v  --version                         Print version\n",
      "\n",
      "Online QUAST manual is available at http://quast.sf.net/manual\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda run -n quast quast --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acc66f85-118e-4cf8-92c1-09070f568cb5",
   "metadata": {},
   "source": [
    "## 5.2. Create a quality report with `quast`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "32e19c56-423d-4226-bb11-0b4b52b6bf57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/CSCScience.ca/apetkau/miniconda3/envs/quast/bin/quast SRR3028792.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-161-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.7.12\n",
      "  CPUs number: 56\n",
      "\n",
      "Started: 2022-01-20 18:22:20\n",
      "\n",
      "Logging to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/quast_results/results_2022_01_20_18_22_20/quast.log\n",
      "NOTICE: Maximum number of threads is set to 14 (use --threads option to set it manually)\n",
      "\n",
      "CWD: /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data\n",
      "Main parameters: \n",
      "  MODE: default, threads: 14, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: one, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  SRR3028792.fasta ==> SRR3028792\n",
      "\n",
      "2022-01-20 18:22:20\n",
      "Running Basic statistics processor...\n",
      "  Contig files: \n",
      "    SRR3028792\n",
      "  Calculating N50 and L50...\n",
      "    SRR3028792, N50 = 298538, L50 = 4, Total length = 4839502, GC % = 52.04, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/quast_results/results_2022_01_20_18_22_20/basic_stats/Nx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/quast_results/results_2022_01_20_18_22_20/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/quast_results/results_2022_01_20_18_22_20/basic_stats/GC_content_plot.pdf\n",
      "  Drawing SRR3028792 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/quast_results/results_2022_01_20_18_22_20/basic_stats/SRR3028792_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2022-01-20 18:22:21\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2022-01-20 18:22:22\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/quast_results/results_2022_01_20_18_22_20/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/quast_results/results_2022_01_20_18_22_20/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/quast_results/results_2022_01_20_18_22_20/report.html\n",
      "  PDF version (tables and plots) is saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/quast_results/results_2022_01_20_18_22_20/report.pdf\n",
      "  Icarus (contig browser) is saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/quast_results/results_2022_01_20_18_22_20/icarus.html\n",
      "  Log is saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/quast_results/results_2022_01_20_18_22_20/quast.log\n",
      "\n",
      "Finished: 2022-01-20 18:22:22\n",
      "Elapsed time: 0:00:02.114222\n",
      "NOTICEs: 2; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n",
      "\n",
      "\u001b[?2004h"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda run -n quast quast SRR3028792.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2ec972d-5049-4529-933a-5b0445ccf568",
   "metadata": {},
   "source": [
    "### 5.2.1 Zip up report for download\n",
    "\n",
    "Let's zip up the quast report for download so we can open it in our browser. You can create a `*.zip` archive from the command-line with the `zip` command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87bd5bcd-7871-43d0-bac7-6fd72cdc5f51",
   "metadata": {},
   "outputs": [],
   "source": [
    "zip -r quast.zip quast_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "828277fb-a6e3-4cc5-8053-183b415ec0e9",
   "metadata": {},
   "source": [
    "# 6. Where to go from here\n",
    "\n",
    "An assembled genome file (`SRR3028792.fasta`) often forms the basis for other downstream analysis. Here are a few examples you can try on your own:\n",
    "\n",
    "1. Try uploading `SRR3028792.fasta` to CARD/RGI (which will identify genes associated with antimicrobial resistance): https://card.mcmaster.ca/analyze/rgi"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
